Service/CredentialsRevoker/DoctrineCredentialsRevoker.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Trikoder\Bundle\OAuth2Bundle\Service\CredentialsRevoker;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Trikoder\Bundle\OAuth2Bundle\Model\AccessToken;
+use Trikoder\Bundle\OAuth2Bundle\Model\AuthorizationCode;
+use Trikoder\Bundle\OAuth2Bundle\Model\Client;
+use Trikoder\Bundle\OAuth2Bundle\Model\RefreshToken;
+use Trikoder\Bundle\OAuth2Bundle\Service\CredentialsRevokerInterface;
+
+final class DoctrineCredentialsRevoker implements CredentialsRevokerInterface
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    public function __construct(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    public function revokeCredentialsForUser(UserInterface $user): void
+    {
+        $userIdentifier = $user->getUsername();
+
+        $this->entityManager->createQueryBuilder()
+            ->update(AccessToken::class, 'at')
+            ->set('at.revoked', true)
+            ->where('at.userIdentifier = :userIdentifier')
+            ->setParameter('userIdentifier', $userIdentifier)
+            ->getQuery()
+            ->execute();
+
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder
+            ->update(RefreshToken::class, 'rt')
+            ->set('rt.revoked', true)
+            ->where(
+                $queryBuilder->expr()->in(
+                    'rt.accessToken',
+                    $this->entityManager->createQueryBuilder()
+                        ->select('at.identifier')
+                        ->from(AccessToken::class, 'at')
+                        ->where('at.userIdentifier = :userIdentifier')
+                        ->getDQL()
+                ))
+            ->setParameter('userIdentifier', $userIdentifier)
+            ->getQuery()
+            ->execute();
+
+        $this->entityManager->createQueryBuilder()
+            ->update(AuthorizationCode::class, 'ac')
+            ->set('ac.revoked', true)
+            ->where('ac.userIdentifier = :userIdentifier')
+            ->setParameter('userIdentifier', $userIdentifier)
+            ->getQuery()
+            ->execute();
+    }
+
+    public function revokeCredentialsForClient(Client $client): void
+    {
+        $doctrineClient = $this->entityManager
+            ->getRepository(Client::class)
+            ->findOneBy(['identifier' => $client->getIdentifier()]);
+
+        $this->entityManager->createQueryBuilder()
+            ->update(AccessToken::class, 'at')
+            ->set('at.revoked', true)
+            ->where('at.client = :client')
+            ->setParameter('client', $doctrineClient)
+            ->getQuery()
+            ->execute();
+
+        $qb = $this->entityManager->createQueryBuilder();
+        $qb->update(RefreshToken::class, 'rt')
+            ->set('rt.revoked', true)
+            ->where($qb->expr()->in(
```

Comment:
@toniperic Could we perhaps use a join here instead of sub-queries?

Reply:
There are two kinda-connected reasons why I've opted for doing a sub-query instead of a join:

1. [This](https://stackoverflow.com/a/2577188/7424875)
2. Some storage drivers that Doctrine supports do not support UPDATEs with JOINs _(eyes at SQLite, but there may be more...)_

Reply:
@toniperic Ok, makes sense.